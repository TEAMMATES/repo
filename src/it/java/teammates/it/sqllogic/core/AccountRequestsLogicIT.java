package teammates.it.sqllogic.core;

import java.time.Instant;
import java.util.UUID;

import org.testng.annotations.Test;

import teammates.common.datatransfer.AccountRequestStatus;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.AccountRequestsLogic;
import teammates.storage.sqlapi.AccountRequestsDb;
import teammates.storage.sqlentity.AccountRequest;

/**
 * SUT: {@link AccountRequestsLogic}.
 */
public class AccountRequestsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private AccountRequestsLogic accountRequestsLogic = AccountRequestsLogic.inst();

    @Test
    public void testGetAccountRequest_nonExistentAccountRequest_returnsNull() {
        UUID id = UUID.randomUUID();
        AccountRequest actualAccountRequest = accountRequestsLogic.getAccountRequest(id);
        assertNull(actualAccountRequest);
    }

    @Test
    public void testGetAccountRequest_existingAccountRequest_getsSuccessfully() throws InvalidParametersException {
        AccountRequest expectedAccountRequest =
                new AccountRequest("test@gmail.com", "name", "institute", AccountRequestStatus.PENDING, "comments");
        UUID id = expectedAccountRequest.getId();
        accountRequestsLogic.createAccountRequest(expectedAccountRequest);
        AccountRequest actualAccountRequest = accountRequestsLogic.getAccountRequest(id);
        assertEquals(expectedAccountRequest, actualAccountRequest);
    }

    @Test
    public void testResetAccountRequest()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {

        ______TS("success: create account request and update registeredAt field");

        String name = "name lee";
        String email = "email@gmail.com";
        String institute = "institute";
        AccountRequestStatus status = AccountRequestStatus.PENDING;
        String comments = "comments";

        AccountRequest toReset = accountRequestsLogic.createAccountRequest(name, email, institute, status, comments);
        AccountRequestsDb accountRequestsDb = AccountRequestsDb.inst();

        toReset.setRegisteredAt(Instant.now());
        toReset = accountRequestsDb.getAccountRequest(email, institute);

        assertNotNull(toReset);
        assertNotNull(toReset.getRegisteredAt());

        ______TS("success: reset account request that already exists");

        AccountRequest resetted = accountRequestsLogic.resetAccountRequest(email, institute);

        assertNull(resetted.getRegisteredAt());

        ______TS("success: test delete account request");

        accountRequestsLogic.deleteAccountRequest(email, institute);

        assertNull(accountRequestsLogic.getAccountRequest(email, institute));

        ______TS("failure: reset account request that does not exist");

        assertThrows(EntityDoesNotExistException.class,
                () -> accountRequestsLogic.resetAccountRequest(name, institute));
    }
}
