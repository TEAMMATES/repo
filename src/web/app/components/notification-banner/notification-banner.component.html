<div class="banner" [ngClass]="notifications[0].style | notificationStyleClass" *ngIf="isShown && notifications.length > 0" @collapseAnim>
  <button id="btn-close-notif" type="button" class="close" aria-label="Close" (click)="closeNotification()">
    <span aria-hidden="true">&times;</span>
  </button>
  <h5>{{ notifications[0].title }}</h5>
  <div class="banner-text" [innerHTML]="notifications[0].message"></div>
  <div class="d-flex flex-row-reverse mt-3">
    <button id="btn-mark-as-read" type="button" [ngClass]="getButtonClass(notifications[0])" (click)="markNotificationAsRead(notifications[0])">
      Mark as Read
    </button>
  </div>
</div>
