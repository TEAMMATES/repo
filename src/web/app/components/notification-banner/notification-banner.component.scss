.banner {
  position: fixed;
  top: 50px;
  width: 100%;
  padding: 20px 100px;
  z-index: 900;
  word-break: break-word;
}

.banner-text {
  max-height: 100px;
  overflow: scroll;
}

@media (max-width: 1024px) {
  .banner {
    padding: 20px 50px;
  }
}

@media (max-width: 768px) {
  .banner {
    padding: 20px;
  }
}
