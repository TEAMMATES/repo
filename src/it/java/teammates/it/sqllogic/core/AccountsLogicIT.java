package teammates.it.sqllogic.core;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

import org.testng.annotations.Test;

import teammates.common.datatransfer.NotificationStyle;
import teammates.common.datatransfer.NotificationTargetUser;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.AccountsLogic;
import teammates.sqllogic.core.NotificationsLogic;
import teammates.storage.sqlapi.AccountsDb;
import teammates.storage.sqlentity.Account;
import teammates.storage.sqlentity.Notification;
import teammates.storage.sqlentity.ReadNotification;

/**
 * SUT: {@link AccountsLogic}.
 */
public class AccountsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private AccountsLogic accountsLogic = AccountsLogic.inst();
    private NotificationsLogic notificationsLogic = NotificationsLogic.inst();

    private AccountsDb accountsDb = AccountsDb.inst();

    @Test
    public void testUpdateReadNotifications()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {
        ______TS("success: mark notification as read");
        Account account = new Account("google-id", "name", "email@teammates.com");
        Notification notification = new Notification(Instant.parse("2011-01-01T00:00:00Z"),
                Instant.parse("2099-01-01T00:00:00Z"), NotificationStyle.DANGER, NotificationTargetUser.GENERAL,
                "A deprecation note", "<p>Deprecation happens in three minutes</p>");
        accountsDb.createAccount(account);
        notificationsLogic.createNotification(notification);

        String googleId = account.getGoogleId();
        UUID notificationId = notification.getId();
        accountsLogic.updateReadNotifications(googleId, notificationId, notification.getEndTime());

        Account actualAccount = accountsDb.getAccountByGoogleId(googleId);
        List<ReadNotification> accountReadNotifications = actualAccount.getReadNotifications();
        assertEquals(1, accountReadNotifications.size());
        assertSame(actualAccount, accountReadNotifications.get(0).getAccount());
        assertSame(notification, accountReadNotifications.get(0).getNotification());
    }
}
