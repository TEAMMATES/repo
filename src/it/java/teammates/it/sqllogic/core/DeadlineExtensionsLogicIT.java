package teammates.it.sqllogic.core;

import java.time.Instant;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import teammates.common.datatransfer.SqlDataBundle;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.common.util.HibernateUtil;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.DeadlineExtensionsLogic;
import teammates.storage.sqlentity.DeadlineExtension;
import teammates.storage.sqlentity.FeedbackSession;
import teammates.storage.sqlentity.Student;

/**
 * SUT: {@link DeadlineExtensionsLogic}.
 */
public class DeadlineExtensionsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private DeadlineExtensionsLogic deadlineExtensionsLogic = DeadlineExtensionsLogic.inst();
    private SqlDataBundle typicalDataBundle;

    @Override
    @BeforeClass
    public void setupClass() {
        super.setupClass();
        typicalDataBundle = getTypicalSqlDataBundle();
    }

    @Override
    @BeforeMethod
    protected void setUp() throws Exception {
        super.setUp();
        persistDataBundle(typicalDataBundle);
        HibernateUtil.flushSession();
        HibernateUtil.clearSession();
    }

    @Test
    public void testGetExtendedDeadline_extensionExists_success() {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student1InCourse1");

        assert student != null;
        Instant extendedDeadlineForStudent = deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student);

        assertNotNull(extendedDeadlineForStudent);
        assertEquals(Instant.parse("2027-04-30T23:00:00Z"), extendedDeadlineForStudent);
    }

    @Test
    public void testGetExtendedDeadline_extensionDoesNotExist_null() {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student2InCourse1");
        Instant extendedDeadlineForStudent = deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student);

        assertNull(extendedDeadlineForStudent);
    }

    @Test
    public void testUpdateDeadlineExtension()
            throws EntityDoesNotExistException, InvalidParametersException {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student1InCourse1");

        DeadlineExtension de = typicalDataBundle.deadlineExtensions.get("student1InCourse1Session1");

        Instant newEndTime = Instant.parse("2028-04-30T23:00:00Z");
        de.setEndTime(newEndTime);
        deadlineExtensionsLogic.updateDeadlineExtension(de);

        assertEquals(newEndTime, deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student));
    }

    @Test
    public void testUpdateDeadlineExtension_invalidParameters_originalUnchanged() {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student1InCourse1");

        DeadlineExtension de = typicalDataBundle.deadlineExtensions.get("student1InCourse1Session1");
        Instant originalEndTime = de.getEndTime();

        Instant invalidEndTime = Instant.parse("2011-04-01T22:00:00Z");
        assert invalidEndTime.isBefore(feedbackSession.getStartTime());
        de.setEndTime(invalidEndTime);

        assertThrows(InvalidParametersException.class, () -> deadlineExtensionsLogic.updateDeadlineExtension(de));

        assertEquals(originalEndTime, deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student));
    }

}
