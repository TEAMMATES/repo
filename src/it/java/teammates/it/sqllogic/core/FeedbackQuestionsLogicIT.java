package teammates.it.sqllogic.core;

import java.util.ArrayList;
import java.util.List;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import teammates.common.datatransfer.FeedbackParticipantType;
import teammates.common.datatransfer.SqlDataBundle;
import teammates.common.datatransfer.questions.FeedbackTextQuestionDetails;
import teammates.common.exception.InvalidParametersException;
import teammates.common.util.HibernateUtil;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.FeedbackQuestionsLogic;
import teammates.storage.sqlentity.FeedbackQuestion;
import teammates.storage.sqlentity.FeedbackSession;

/**
 * SUT: {@link FeedbackQuestionsLogic}.
 */
public class FeedbackQuestionsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private FeedbackQuestionsLogic fqLogic = FeedbackQuestionsLogic.inst();

    private SqlDataBundle typicalDataBundle;

    @Override
    @BeforeClass
    public void setupClass() {
        super.setupClass();
        typicalDataBundle = getTypicalSqlDataBundle();
    }

    @Override
    @BeforeMethod
    protected void setUp() throws Exception {
        super.setUp();
        persistDataBundle(typicalDataBundle);
        HibernateUtil.flushSession();
    }

    @Test
    public void testCreateFeedbackQuestion() throws InvalidParametersException {
        FeedbackSession fs = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        FeedbackTextQuestionDetails newQuestionDetails = new FeedbackTextQuestionDetails("New question text.");
        List<FeedbackParticipantType> showTos = new ArrayList<>();
        showTos.add(FeedbackParticipantType.INSTRUCTORS);
        FeedbackQuestion newQuestion = FeedbackQuestion.makeQuestion(fs, 6, "This is a new text question",
                FeedbackParticipantType.STUDENTS, FeedbackParticipantType.OWN_TEAM_MEMBERS, -100,
                showTos, showTos, showTos, newQuestionDetails);

        newQuestion = fqLogic.createFeedbackQuestion(newQuestion);

        FeedbackQuestion actualQuestion = fqLogic.getFeedbackQuestion(newQuestion.getId());

        verifyEquals(newQuestion, actualQuestion);
    }

    @Test
    public void testGetFeedbackQuestionsForSession() {
        FeedbackSession fs = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        FeedbackQuestion fq1 = typicalDataBundle.feedbackQuestions.get("qn1InSession1InCourse1");
        FeedbackQuestion fq2 = typicalDataBundle.feedbackQuestions.get("qn2InSession1InCourse1");
        FeedbackQuestion fq3 = typicalDataBundle.feedbackQuestions.get("qn3InSession1InCourse1");
        FeedbackQuestion fq4 = typicalDataBundle.feedbackQuestions.get("qn4InSession1InCourse1");
        FeedbackQuestion fq5 = typicalDataBundle.feedbackQuestions.get("qn5InSession1InCourse1");
        FeedbackQuestion fq6 = typicalDataBundle.feedbackQuestions.get("qn6InSession1InCourse1NoResponses");

        List<FeedbackQuestion> expectedQuestions = List.of(fq1, fq2, fq3, fq4, fq5, fq6);

        List<FeedbackQuestion> actualQuestions = fqLogic.getFeedbackQuestionsForSession(fs);

        assertEquals(expectedQuestions.size(), actualQuestions.size());
        assertTrue(expectedQuestions.containsAll(actualQuestions));
    }

}
