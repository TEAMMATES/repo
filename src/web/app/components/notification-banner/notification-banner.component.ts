import { Component, Input, OnInit } from '@angular/core';
import { NotificationService } from '../../../services/notification.service';
import { Notification, Notifications, NotificationTargetUser } from '../../../types/api-output';
import { collapseAnim } from '../teammates-common/collapse-anim';

/**
 * Banner used to display notifications to the user.
 */
@Component({
  selector: 'tm-notification-banner',
  templateUrl: './notification-banner.component.html',
  styleUrls: ['./notification-banner.component.scss'],
  animations: [collapseAnim],
})
export class NotificationBannerComponent implements OnInit {

  @Input()
  notificationTargetUser: NotificationTargetUser = NotificationTargetUser.GENERAL;

  isShown: boolean = true;
  notifications: Notification[] = [];

  constructor(private notificationService: NotificationService) { }

  ngOnInit(): void {
    if (this.notificationTargetUser !== NotificationTargetUser.GENERAL) {
      this.fetchNotifications();
    }
  }

  fetchNotifications(): void {
    this.notificationService.getUnreadNotificationsForTargetUser(this.notificationTargetUser)
      .subscribe((response: Notifications) => {
        this.notifications = response.notifications;
      });
  }

  closeNotification(): void {
    this.isShown = false;
  }
}
