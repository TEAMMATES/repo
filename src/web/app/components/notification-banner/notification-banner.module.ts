import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NotificationBannerComponent } from './notification-banner.component';

/**
 * Module for banner used to display notifications to the user.
 */
@NgModule({
  declarations: [NotificationBannerComponent],
  exports: [NotificationBannerComponent],
  imports: [CommonModule],
})
export class NotificationBannerModule { }
