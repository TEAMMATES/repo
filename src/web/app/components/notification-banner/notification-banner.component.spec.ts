import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';

import { NotificationBannerComponent } from './notification-banner.component';

describe('NotificationBannerComponent', () => {
  let component: NotificationBannerComponent;
  let fixture: ComponentFixture<NotificationBannerComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [NotificationBannerComponent],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(NotificationBannerComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});
