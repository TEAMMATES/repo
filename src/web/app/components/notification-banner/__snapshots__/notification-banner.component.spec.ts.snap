// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationBannerComponent should snap with 1 unread notification 1`] = `
<tm-notification-banner
  isShown={[Function Boolean]}
  notificationService={[Function NotificationService]}
  notificationTargetUser={[Function String]}
  notifications={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  url=""
>
  <div
    class="banner alert alert-success ng-trigger ng-trigger-collapseAnim ng-animating"
    id="notification-banner"
    style=""
  >
    <button
      aria-label="Close"
      class="close"
      id="btn-close-notif"
      type="button"
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </button>
    <h5>
      valid title 1
    </h5>
    <div
      class="banner-text"
    >
      valid message 1
    </div>
    <div
      class="d-flex flex-row-reverse mt-3 mr-2"
    >
      <button
        class="btn btn-success"
        id="btn-mark-as-read"
        type="button"
      >
         Mark as Read 
      </button>
    </div>
  </div>
</tm-notification-banner>
`;

exports[`NotificationBannerComponent should snap with multiple unread notifications 1`] = `
<tm-notification-banner
  isShown={[Function Boolean]}
  notificationService={[Function NotificationService]}
  notificationTargetUser={[Function String]}
  notifications={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  url=""
>
  <div
    class="banner alert alert-success ng-trigger ng-trigger-collapseAnim ng-animating"
    id="notification-banner"
    style=""
  >
    <button
      aria-label="Close"
      class="close"
      id="btn-close-notif"
      type="button"
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </button>
    <h5>
      valid title 1
    </h5>
    <div
      class="banner-text"
    >
      valid message 1
    </div>
    <div
      class="d-flex flex-row-reverse mt-3 mr-2"
    >
      <button
        class="btn btn-success"
        id="btn-mark-as-read"
        type="button"
      >
         Mark as Read 
      </button>
    </div>
  </div>
</tm-notification-banner>
`;

exports[`NotificationBannerComponent should snap with no unread notifications 1`] = `
<tm-notification-banner
  isShown={[Function Boolean]}
  notificationService={[Function NotificationService]}
  notificationTargetUser={[Function String]}
  notifications={[Function Array]}
  statusMessageService={[Function StatusMessageService]}
  url=""
>
</tm-notification-banner>
`;
